movement/lib/TOTP/sha512.c

```diff
@@ -0,0 +1,422 @@
+    mbedtls_sha512_update( ctx, msglen, 16 );
+
+    PUT_UINT64_BE( ctx->state[0], output,  0 );
+    PUT_UINT64_BE( ctx->state[1], output,  8 );
+    PUT_UINT64_BE( ctx->state[2], output, 16 );
+    PUT_UINT64_BE( ctx->state[3], output, 24 );
+    PUT_UINT64_BE( ctx->state[4], output, 32 );
+    PUT_UINT64_BE( ctx->state[5], output, 40 );
+
+    if( ctx->is384 == 0 )
+    {
+        PUT_UINT64_BE( ctx->state[6], output, 48 );
+        PUT_UINT64_BE( ctx->state[7], output, 56 );
+    }
+}
+
+/*
+ * output = SHA-512( input buffer )
+ */
+void mbedtls_sha512( const unsigned char *input, size_t ilen,
+             unsigned char* output, int is384 )
+{
+    mbedtls_sha512_context ctx;
+
+    mbedtls_sha512_init( &ctx );
+    mbedtls_sha512_starts( &ctx, is384 );
+    mbedtls_sha512_update( &ctx, input, ilen );
+    mbedtls_sha512_finish( &ctx, output );
+    mbedtls_sha512_free( &ctx );
+}
+
+/*
+* Compute HMAC_SHA384/512 using key, key length, text to hash, size of the text, output buffer and a switch for SHA384
+*/
+void HMAC_SHA512(const uint8_t* key, size_t key_length, const uint8_t *in, size_t n, uint8_t* out, int is384){
+  int digest_length = SHA512_DIGEST_LENGTH;
+  if (is384 == 1) {
+    digest_length = SHA384_DIGEST_LENGTH;
+  }
+
+  uint8_t i;
+  uint8_t k_ipad[SHA512_BLOCK_LENGTH]; /* inner padding - key XORd with ipad */
+  uint8_t k_opad[SHA512_BLOCK_LENGTH]; /* outer padding - key XORd with opad */
+  uint8_t buffer[SHA512_BLOCK_LENGTH + digest_length];
+
+  /* start out by storing key in pads */
+  memset(k_ipad, 0, sizeof(k_ipad));
+  memset(k_opad, 0, sizeof(k_opad));
+
+  if (key_length <= SHA512_BLOCK_LENGTH) {
+      memcpy(k_ipad, key, key_length);
+      memcpy(k_opad, key, key_length);
+  }
+
+  else {
+      mbedtls_sha512(key, key_length, k_ipad, is384);
+      memcpy(k_opad, k_ipad, SHA512_BLOCK_LENGTH);
+  }
+
+  /* XOR key with ipad and opad values */
+  for (i = 0; i < SHA512_BLOCK_LENGTH; i++) {
+      k_ipad[i] ^= HMAC_IPAD;
+      k_opad[i] ^= HMAC_OPAD;
+  }
+  
+  // perform inner SHA512
+  memcpy(buffer, k_ipad, SHA512_BLOCK_LENGTH);
+  memcpy(buffer + SHA512_BLOCK_LENGTH, in, n);
+  mbedtls_sha512(buffer, SHA512_BLOCK_LENGTH + n, out, is384);
+  
+  memset(buffer, 0, SHA512_BLOCK_LENGTH + n);
+
+  // perform outer SHA512
+  memcpy(buffer, k_opad, SHA512_BLOCK_LENGTH);
+  memcpy(buffer + SHA512_BLOCK_LENGTH, out, digest_length);
+  mbedtls_sha512(buffer, SHA512_BLOCK_LENGTH + digest_length, out, is384);
+}
+
+/*
+* Compute TOTP_HMAC_SHA384/512 using key, key length, text to hash, size of the text and a switch for SHA384
+*/
+uint32_t TOTP_HMAC_SHA512(const uint8_t* key, size_t key_length, const uint8_t *in, size_t n, int is384){
+    int digest_length = SHA512_DIGEST_LENGTH;
+    if (is384 == 1) {
+      digest_length = SHA384_DIGEST_LENGTH;
+    }
+
+    // STEP 1, get the HMAC-SHA512 hash from counter and key
+    uint8_t hash[digest_length];
+    HMAC_SHA512(key, key_length, in, 8, hash, is384);

```

Comment:
Same issue as in TOTP_HMAC_SHA1.

Reply:
You are right :). This was fixed in https://github.com/joeycastillo/Sensor-Watch/pull/151/commits/db2b82a5dc0e310489e966d1cff902568627d642.